pkg/paho-mqtt/contrib/riot_iface.c

```diff
@@ -0,0 +1,237 @@
+#define PAHO_MQTT_YIELD_MS                  (10)
+#endif
+
+static int mqtt_read(struct Network *n, unsigned char *buf, int len,
+                     int timeout_ms)
+{
+    int _timeout;
+    int _len;
+    void *temp_buf;
+    int rc = -1;
+
+#ifdef MODULE_LWIP
+    /* As LWIP doesn't support reads packets byte per byte and
+     * PAHO MQTT reads like that to decode it on the fly,
+     * we read TSRB_MAX_SIZE at once and keep them in a ring buffer.
+     */
+    uint8_t _internal_buffer[TSRB_MAX_SIZE];
+    temp_buf = _internal_buffer;
+    _len = TSRB_MAX_SIZE;
+    _timeout = 0;
+#else
+    temp_buf = buf;
+    _len = len;
+    _timeout = timeout_ms;
+#endif
+
+    uint64_t send_tick = xtimer_now64().ticks64 + xtimer_ticks_from_usec64(timeout_ms * US_PER_MS).ticks64;
+    do {
+        rc = sock_tcp_read(&n->sock, temp_buf, _len, _timeout);
+        if (rc == -EAGAIN) {
+            rc = 0;
+        }
+
+    #ifdef MODULE_LWIP
+        if (rc > 0) {
+            tsrb_add(&tsrb_lwip_tcp, temp_buf, rc);
+        }
+
+        rc = tsrb_get(&tsrb_lwip_tcp, buf, len);
+    #endif
+    } while (rc < len && xtimer_now64().ticks64 < send_tick && rc >= 0);
+
+#ifdef ENABLE_DEBUG
+#ifdef MODULE_LWIP
+    if (rc > 0) {
+        DEBUG("MQTT buf asked for %d, available to read %d\n", rc, tsrb_avail(&tsrb_lwip_tcp));
+        for (int i = 0; i < rc; i++) {
+            DEBUG("0x%02X ", buf[i]);
+        }
+        DEBUG("\n");
+    }
+#endif
+#endif
+
+    return rc;
+}
+
+static int mqtt_write(struct Network *n, unsigned char *buf, int len,
+                      int timeout_ms)
+{
+    /* timeout is controlled by upper layer in PAHO */
+    (void) timeout_ms;
+
+    return sock_tcp_write(&n->sock, buf, len);
+}
+
+void NetworkInit(Network *n)
+{
+#ifdef MODULE_LWIP
+    tsrb_init(&tsrb_lwip_tcp, buffer, TSRB_MAX_SIZE);
+#endif
+    n->mqttread = mqtt_read;
+    n->mqttwrite = mqtt_write;
+}
+
+int NetworkConnect(Network *n, char *addr_ip, int port)
+{
+    int ret =-1;
+    sock_tcp_ep_t *remote = NULL;
+    char _local_ip[IP_MAX_LEN_ADDRESS];
+    /* ipvN_addr_from_str modifies input buffer */
+    strncpy(_local_ip, addr_ip, sizeof(_local_ip));
+
+
+    sock_tcp_ep_t remote4 = SOCK_IPV4_EP_ANY;
+    if (ipv4_addr_from_str((ipv4_addr_t *)&remote4.addr,
+                           _local_ip) == NULL) {
+#ifdef SOCK_HAS_IPV6
+        LOG_INFO("Error IPv4: unable to parse destination address, trying with IPv6\n");
+        sock_tcp_ep_t remote6 = SOCK_IPV6_EP_ANY;
+        if (ipv6_addr_from_str((ipv6_addr_t *)&remote6.addr, _local_ip) == NULL) {
+            LOG_ERROR("Error IPv6: unable to parse destination address\n");
+            return ret;
+        }
+        remote6.port = port;
+        remote = &remote6;
+        goto connect;
+#else
+        LOG_ERROR("Error IPv4: unable to parse destination address\n");
+#endif
+        return ret;
+    }
+    remote4.port = port;
+    remote = &remote4;
+#ifdef SOCK_HAS_IPV6
+connect:
+#endif
```

Comment:
Although this part of the code requires further changes to solve the problem with the undefined `ipv4_addr_from_str`, a first suggested change to point out the following problems:

1. Don't define a variable (`remote6`) inside a local scope and use the address of it outside the scope. In optimized code the lifetime of the variable is only guaranteed inside the scope. This led to an assertion for IPv6 on ESP32.
2. You should use gotos only if it is better for the structure. In this case, the structure without the goto is much more clear. If `ipv4_addr_from_str` fails you try to set `remote6`, otherwise you set `remote4`. The goto makes it less readable than an `if ... else`.

```suggestion
    sock_tcp_ep_t remote4 = SOCK_IPV4_EP_ANY;
    sock_tcp_ep_t remote6 = SOCK_IPV6_EP_ANY;
    
    if (ipv4_addr_from_str((ipv4_addr_t *)&remote4.addr,
                           _local_ip) == NULL) {
#ifdef SOCK_HAS_IPV6
        LOG_INFO("Error IPv4: unable to parse destination address, trying with IPv6\n");
        if (ipv6_addr_from_str((ipv6_addr_t *)&remote6.addr, _local_ip) == NULL) {
            LOG_ERROR("Error IPv6: unable to parse destination address\n");
            return ret;
        }
        remote6.port = port;
        remote = &remote6;
#else
        LOG_ERROR("Error IPv4: unable to parse destination address\n");
#endif
        return ret;
    }
    else {
        remote4.port = port;
        remote = &remote4;
    }
```

Reply:
I' ve changed the implementation as with all the `#ifdef` was becoming messy... but now I don't know what flag to use because there' s no `SOCK_HAS_IPV4` flag

Reply:
```
if (IS_USED(MODULE_LWIP_SOCK_TCP) && IS_USED(MODULE_LWIP_IPV4)) {
}
```